selene_sdk/samplers/dataloader.py

```diff
@@ -0,0 +1,308 @@
+    batch_size : int
+        The number of samples the iterator returns in one step.
+
+    """
+    def __init__(self,
+                 sampler,
+                 num_workers=1,
+                 batch_size=1,
+                 seed=436):
+        def worker_init_fn(worker_id):
+            """
+            This function is called to initialize each worker with different
+            numpy seeds (torch seeds are set by DataLoader automatically).
+            """
+            np.random.seed(seed + worker_id)
+
+        args = {
+            "batch_size": batch_size,
+            "num_workers": num_workers,
+            "pin_memory": True,
+            "worker_init_fn": worker_init_fn
+        }
+
+        super(SamplerDataLoader, self).__init__(_SamplerDataset(sampler), **args)
+        self.seed = seed
+
+
+class _H5Dataset(Dataset):
+    """
+    This class provides a Dataset that directly loads sequences and targets
+    from a hdf5 file. `_H5Dataset` is intended to be used internally by
+    `H5DataLoader`.
+
+    Parameters
+    ----------
+    file_path : str
+        The file path of the hdf5 file.
+    size : int or None
+        Default is None. Specify dataset size. This can be used to limit the
+        dataset to first `size` rows. If None, use the sequences array length.
+    in_memory : bool, optional
+        Default is False. If True, load entire dataset into memory.
+    unpackbits : bool, optional
+        Default is False. If True, unpack binary-valued array from uint8
+        sequence and targets array. See `numpy.packbits` for details.
+    seq_key : str, optional
+        Default is "sequences". Specify the name of the hdf5 dataset that contains
+        sequence data.
+    tgt_key : str, optional
+        Default is "targets". Specify the name of the hdf5 dataset that contains
+        target data.
+
+    Attributes
+    ----------
+    file_path : str
+        The file path of the hdf5 file.
+    size : int
+        Dataset size.
+    in_memory : bool
+        If True, load entire dataset into memory.
+    unpackbits : bool
+        If True, unpack binary-valued array from uint8
+        sequence and targets array. See `numpy.packbits` for details.
+    """
+    def __init__(self,
+                 file_path,
+                 size=None,
+                 in_memory=False,
+                 unpackbits=False,
+                 seq_key="sequences",
+                 tgt_key="targets"):
+        super(_H5Dataset, self).__init__()
+        self.file_path = file_path
+        self.in_memory = in_memory
+        self.unpackbits = unpackbits
+        self.size = size
+
+        self._initialized = False
+        self._seq_key = seq_key
+        self._tgt_key = tgt_key
+
+    def init(func):
+        # delay initialization to allow multiprocessing
+        def dfunc(self, *args, **kwargs):
+            if not self._initialized:
+                self.db = h5py.File(self.file_path, 'r')
```

Comment:
need to document so I want to confirm this - what are the requirements for using the H5Dataset class? (must have sequences, targets, and sequences_length, targets_length attributes?) Where does it get used usually? (MatFileSampler for test and validation?) 

We have an option for packbits that we need to release a script for

Reply:
This can be used with MultiFileSampler similar to SamplerDataLoader. 

This class takes `seq_key` and `tgt_key` arguments and the h5 file need to have two arrays matching these names (I am pretty sure you wrote this haha - I migrated it from the multibin branch since I think it can be useful).  So it needs a sequence array with the name  `seq_key`  and a target array named `tgt_key`  and the sequence length and target dimensions `{seq_key}_length` and `{tgt_key}_length`. The length is needed to support unpackbits.

Reply:
I have updated the docs to add instructions for the input and make `{seq_key}_length` and `{tgt_key}_length` optional (only when unpackbits is needed).  We had a script for generating hdf5 file for multibin - I will look into whether we can include it in some form.

Reply:
I haven't committed it yet - will do it after I tested.

Reply:
are you working on this? if not I'm also happy to do this, I'd like to add it as a script to the top-level of the repo (like where `selene_cli.py` used to be or something - what do you think?

Reply:
I actually just added the docs for H5DataLoader about the format needed but not the script yet. I just looked into the script, you actually wrote a script `write_multibin_mats.py` before that does this and we've added the packbits option. I tried it and it still works. We can probably just add that script as `write_samples_h5.py` right? Not sure about placement - should it be on the top level or should we add a scripts directory?

Reply:
yes the renamed file sounds good! and we can add a scripts directory sure - will you update the PR with it? thanks so much :)

Reply:
sounds good! yes I have added the script and referred to it in the H5DataLoader doc